pysrc/bytewax/execution.py

```diff
@@ -1,20 +1,113 @@
-from typing import Any, Callable, Dict, Iterable, List, Optional, Tuple
+import asyncio
+import threading
+import time
+from collections import abc, deque
+from concurrent.futures import as_completed, ThreadPoolExecutor
+from queue import Empty, Full
 
-from multiprocess import Manager, Pool
+from typing import Any, Callable, Iterable, Tuple
 
-from .bytewax import *  # This will import PyO3 module contents.
+from multiprocess import Manager, Pool, TimeoutError
 
+from .bytewax import Dataflow, WorkerBuilder, WorkerCoro
 
-def run(flow: Dataflow, inp: Iterable[Tuple[int, Any]]) -> List[Tuple[int, Any]]:
-    """Pass data through a dataflow running in the current thread.
 
-    Blocks until execution is complete.
+class WorkerCoroWrapper(abc.Coroutine, abc.Iterable):
+    """We can't subclass Python-native classes in PyO3 yet.
+
+    This wrapper class implements abc.Coroutine and abc.Iterable since
+    we need that for asyncio compatibility.
+
+    https://github.com/PyO3/pyo3/issues/991
+
+    """
+
+    def __init__(self, wrap: WorkerCoro):
+        self.wrap = wrap
+
+    def send(self, value):
+        return self.wrap.send(value)
+
+    def close(self):
+        return self.wrap.close()
+
+    def throw(self, type, value, traceback):
+        return self.wrap.throw(type, value, traceback)
+
+    def __await__(self):
+        return self.wrap.__await__()
+
+    def __next__(self):
+        return self.wrap.__next__()
+
+    def __iter__(self):
+        return self.wrap.__iter__()
+
+
+async def run_main(
+    flow: Dataflow,
+    input_builder: Callable[[int, int], Iterable[Tuple[int, Any]]],
+    output_builder: Callable[[int, int], Callable[[Tuple[int, Any]], None]],
+    should_stop: Callable[[], bool] = lambda: False,
+) -> None:
+    """Execute a dataflow in the current thread.
+
+    You'd commonly use this for prototyping custom input and output
+    builders with a single worker before using them in a cluster
+    setting.
+
+    Returns after execution is complete. Use with `asyncio.run()` if
```

Comment:
What does this mean to "use with asyncio.run()" ? Perhaps an example or direction to docs?

Reply:
Yes, can put an example of that. It'd look like `asyncio.run(run_main(...))`.